hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java

```diff
@@ -998,9 +996,9 @@ public void runMetaSync() {
    * this constraint.
    */
   private void setupWriteClient(Option<JavaRDD<HoodieRecord>> recordsOpt) throws IOException {
-    if ((null != schemaProvider)) {
+    if (null != schemaProvider) {
       Schema sourceSchema = schemaProvider.getSourceSchema();
-      Schema targetSchema = schemaProvider.getTargetSchema();
+      Schema targetSchema = getSchemaForWriteConfig(schemaProvider.getTargetSchema());
```

Comment:
we call reInitWriteClient twice in this class. If we wer fixing the target schema here to go through getSchemaForWriteConfig, should we fix the other caller as well. 
Also, looks like we are calling getSchemaForWriteConfig within reInitWriteClient(), should we remove that call if callers already take care of it.

Reply:
We don't call `getSchemaForWriteConfig` from `reInitWriteClient()`. There is a call to `getHoodieClientConfig` that takes in a schema that must be non-null to hit the `getSchemaForWriteConfig` path though. 

I think one way to simplify this is to make this call always happen if the schema is required in the config. When we initialize the StreamSync class, we will pass in some schema or null potentially but I don't think we really even need the writer schema for this case and want to avoid a read from the filesystem if possible to fetch the latest commit schema. If you agree, I can clean this up so that `getHoodieClientConfig` takes in a schema and a boolean indicating whether the schema must be set in the config.

Reply:
We do call. 
```
reInitWriteClient {

       final HoodieWriteConfig initialWriteConfig = getHoodieClientConfig(targetSchema); 
}
```

within getHoodieClientConfig, we do call 
```
    if (schema != null) {
      builder.withSchema(getSchemaForWriteConfig(schema).toString());
    }
```
  
but I agree, that if we had already called getSchemaForWriteConfig and fixed the target schema, it will be no-op. Was just pointing out we are making repeated calls.

Reply:
@nsivabalan can you link the line where we make the call? I am missing something or have some outdated code somehow.

Reply:
https://github.com/apache/hudi/blob/50f0d9f3baeafe92c38dfb0b59b337d83391ea42/hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java#L1008 

https://github.com/apache/hudi/blob/50f0d9f3baeafe92c38dfb0b59b337d83391ea42/hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java#L1049 

https://github.com/apache/hudi/blob/50f0d9f3baeafe92c38dfb0b59b337d83391ea42/hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java#L1076C26-L1076C49

https://github.com/apache/hudi/blob/50f0d9f3baeafe92c38dfb0b59b337d83391ea42/hudi-utilities/src/main/java/org/apache/hudi/utilities/streamer/StreamSync.java#L1111

Reply:
`getHoodieClientConfig` is also called from the constructor so you need to consider that path too if you are looking simply at paths that can eventually hit this `getSchemaForWriteConfig` method. I've updated the code so that there is no repeated call anymore and the call from the constructor avoids a potential schema lookup entirely.